### PR TITLE
#356 - add benchmarks file

### PR DESCRIPTION
**Pull request by mforets**
See #356.

### CODE REVIEW

**Comment by mforets** [comment]
# Benchmark Report for *LazySets*

## Job Properties
* Time of benchmark: 29 May 2018 - 21:14
* Package commit: e4d002
* Julia commit: 0d7248
* Julia command flags: None
* Environment variables: None

## Results
Below is a table of this job's results, obtained by running the benchmarks.
The values listed in the `ID` column have the structure `[parent_group, child_group, ..., key]`, and can be used to
index into the BaseBenchmarks suite to retrieve the corresponding benchmarks.
The percentages accompanying time and memory values in the below table are noise tolerances. The "true"
time/memory value for a given benchmark is expected to fall within this percentage of the reported value.
An empty cell means that the value was zero.

| ID                                               | time            | GC time | memory         | allocations |
|--------------------------------------------------|----------------:|--------:|---------------:|------------:|
| `["Balls", "(\"LazySets.Ball1\", 1)"]`           |  96.000 ns (5%) |         | 192 bytes (1%) |           2 |
| `["Balls", "(\"LazySets.Ball1\", 10)"]`          | 122.000 ns (5%) |         | 320 bytes (1%) |           2 |
| `["Balls", "(\"LazySets.Ball1\", 100)"]`         | 323.000 ns (5%) |         |  1.75 KiB (1%) |           2 |
| `["Balls", "(\"LazySets.Ball1\", 1000)"]`        |   2.884 μs (5%) |         | 15.88 KiB (1%) |           2 |
| `["Balls", "(\"LazySets.Ball2\", 1)"]`           |  85.000 ns (5%) |         |  96 bytes (1%) |           1 |
| `["Balls", "(\"LazySets.Ball2\", 10)"]`          | 102.000 ns (5%) |         | 160 bytes (1%) |           1 |
| `["Balls", "(\"LazySets.Ball2\", 100)"]`         | 144.000 ns (5%) |         | 896 bytes (1%) |           1 |
| `["Balls", "(\"LazySets.Ball2\", 1000)"]`        | 838.000 ns (5%) |         |  7.94 KiB (1%) |           1 |
| `["Balls", "(\"LazySets.BallInf\", 1)"]`         | 100.000 ns (5%) |         | 192 bytes (1%) |           2 |
| `["Balls", "(\"LazySets.BallInf\", 10)"]`        | 140.000 ns (5%) |         | 320 bytes (1%) |           2 |
| `["Balls", "(\"LazySets.BallInf\", 100)"]`       | 517.000 ns (5%) |         |  1.75 KiB (1%) |           2 |
| `["Balls", "(\"LazySets.BallInf\", 1000)"]`      |   4.426 μs (5%) |         | 15.88 KiB (1%) |           2 |
| `["Linear map", "(\"LazySets.Ball1\", 1)"]`      | 400.000 ns (5%) |         | 384 bytes (1%) |           4 |
| `["Linear map", "(\"LazySets.Ball1\", 10)"]`     | 554.000 ns (5%) |         | 640 bytes (1%) |           4 |
| `["Linear map", "(\"LazySets.Ball1\", 100)"]`    |   7.352 μs (5%) |         |  3.50 KiB (1%) |           4 |
| `["Linear map", "(\"LazySets.Ball1\", 1000)"]`   | 246.566 μs (5%) |         | 31.75 KiB (1%) |           4 |
| `["Linear map", "(\"LazySets.Ball2\", 1)"]`      | 381.000 ns (5%) |         | 288 bytes (1%) |           3 |
| `["Linear map", "(\"LazySets.Ball2\", 10)"]`     | 546.000 ns (5%) |         | 480 bytes (1%) |           3 |
| `["Linear map", "(\"LazySets.Ball2\", 100)"]`    |   7.253 μs (5%) |         |  2.63 KiB (1%) |           3 |
| `["Linear map", "(\"LazySets.Ball2\", 1000)"]`   | 244.305 μs (5%) |         | 23.81 KiB (1%) |           3 |
| `["Linear map", "(\"LazySets.BallInf\", 1)"]`    | 401.000 ns (5%) |         | 384 bytes (1%) |           4 |
| `["Linear map", "(\"LazySets.BallInf\", 10)"]`   | 570.000 ns (5%) |         | 640 bytes (1%) |           4 |
| `["Linear map", "(\"LazySets.BallInf\", 100)"]`  |   7.598 μs (5%) |         |  3.50 KiB (1%) |           4 |
| `["Linear map", "(\"LazySets.BallInf\", 1000)"]` | 248.361 μs (5%) |         | 31.75 KiB (1%) |           4 |

## Benchmark Group List
Here's a list of all the benchmark groups executed by this job:

- `["Balls"]`
- `["Linear map"]`

## Julia versioninfo
```
Julia Version 0.6.1
Commit 0d7248e (2017-10-24 22:15 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin14.5.0)
  CPU: Intel(R) Core(TM) i7-4770HQ CPU @ 2.20GHz
  WORD_SIZE: 64
  uname: Darwin 15.6.0 Darwin Kernel Version 15.6.0: Wed Nov  2 20:30:56 PDT 2016; root:xnu-3248.60.11.1.2~2/RELEASE_X86_64 x86_64 i386
Memory: 16.0 GB (313.65625 MB free)
Uptime: 381213.0 sec
Load Avg:  4.15478515625  3.314453125  3.1357421875
Intel(R) Core(TM) i7-4770HQ CPU @ 2.20GHz: 
       speed         user         nice          sys         idle          irq
#1  2200 MHz     224133 s          0 s     181363 s    1306402 s          0 s
#2  2200 MHz      17547 s          0 s      15591 s    1678599 s          0 s
#3  2200 MHz     216780 s          0 s      99923 s    1395119 s          0 s
#4  2200 MHz      18577 s          0 s      15619 s    1677515 s          0 s
#5  2200 MHz     217236 s          0 s     100133 s    1394422 s          0 s
#6  2200 MHz      18646 s          0 s      15677 s    1677365 s          0 s
#7  2200 MHz     217285 s          0 s      99952 s    1394524 s          0 s
#8  2200 MHz      18704 s          0 s      15847 s    1677119 s          0 s

  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.9.1 (ORCJIT, haswell)

File: benchmark/benchmarks.jl [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+julia> results = benchmarkpkg("LazySets")
+```
+
+To compare current version to another tagged version:
+
+```julia
+julia> results = judge("LazySets", "v1.2.0")
+```
+
+To export the benchmark results to a Markwodnfile:
```
Comment (schillic) [inline benchmark/benchmarks.jl]: Markdown file

**Comment by schillic** [comment]
> Would it be better to have separate files for separate sub-groups? I imagine that in the long run we want to have this for all set types, etc.

In particular, I think it will be helpful to comment most benchmarks and only create a single one when you try to tweak performance. I often do this for `runtests.jl`, so this file here could just have a list of includes in the same spirit.

File: benchmark/benchmarks.jl [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+```julia
+julia> using PkgBenchmark
+julia> results = benchmarkpkg("LazySets")
+```
+
+To compare current version to another tagged version:
+
+```julia
+julia> results = judge("LazySets", "v1.2.0")
```
Comment (schillic) [inline benchmark/benchmarks.jl]: This crashes because in that version we did not have a `benchmarks.jl` file.

Comment (mforets) [reply benchmark/benchmarks.jl]: true, thanks, i changed it to `results = judge("LazySets", <tagged-version-or-branch>)`

**Comment by mforets** [comment]
> Would it be better to have separate files for separate sub-groups?

Yes, i've refactored a bit. In another PR we could add `overapproximate` for example.

> In particular, I think it will be helpful to comment most benchmarks and only create a single one when you try to tweak performance.

👍 
BTW `benchmarkpkg`  i expect to be a configuration to only do specific benchmark groups (actually `SUITE` is just array of benchmark groups), but [in the docs](https://juliaci.github.io/PkgBenchmark.jl/latest/run_benchmarks.html) i didn't find it.

**Comment by schillic** [comment]
> BTW benchmarkpkg i expect to be a configuration to only do specific benchmark groups (actually SUITE is just array of benchmark groups), but in the docs i didn't find it.

I do not think so, because how should it know which lines of the code to execute? I think this is only for collecting the results and presenting them in the end.

**Comment by mforets** [comment]
for example to ask that from `SUITE` i want to benchmark only the `SUITE["Balls"]` benchmark subgroup.

**Comment by schillic** [comment]
> for example to ask that from SUITE i want to benchmark only the SUITE["Balls"] benchmark subgroup.

Yes, but how should the code know what it has to do then? Only executing the lines that write to the "Balls" key is clearly not working.

File: benchmark/benchmarks.jl (hunk 1 above)
Comment (schillic) [reply benchmark/benchmarks.jl]: My main point was the typo in "Markdown", which is still there :smile:

**Comment by mforets** [comment]
.. i though that passing the name of a specific benchmark (sub)-group it would only do that one, for us like `results = benchmarkpkg("LazySets", "Balls")`, but this doesn't work.

**Comment by mforets** [comment]
the final touch would be to transform these tables into plots.

in the time column, do you know  what is exactly the percentage in parentheses? `7.598 μs (5%)` i suppose is some measure of dispersion around the mean value..

**Comment by schillic** [comment]
I also thought so, but I do not know.

**Pull request merged**